File: week7/q2.java
// Alternating Add and Subtract (Curried
// Function)
//approach : first create a constructor then create a function that takes multiple apply() as parameters then by switching a boolean each time
// time complexity : O(n)
// space complexity : O(1)
public class AlternatingSum {
    private final int v;
    private final boolean add;

    public AlternatingSum(int v, boolean add) {
        this.v = v;
        this.add = add;
    }
    public Function<Integer, AlternatingSum> apply() {
        return (Integer x) -> {
            if (addNext) {
                return new AlternatingSum(value + x, false);
            } else {
                return new AlternatingSum(value - x, true);
            }
        };
    }
    public static AlternatingSum add_subtract(int num) {
        return new AlternatingSum(num, false);
    }
    
    public static void main(String[] args) {
        System.out.println(add_subtract(7)); 
        System.out.println(add_subtract(1).apply().apply(2).apply(3)); 
        System.out.println(add_subtract(-5).apply().apply(10).apply(3).apply(9));
    }
}

File: week7/q2.cpp
//Alternating Add and Subtract without curied function
 // same approach by seitching a boolean variable each time
// time compelxity : O(n)
// space complexity : O(1)

#include <bits/stdc++.h>
using namespace std;

int main(){
    int n;
    cin >> n;
    int sum=0;
    int z;
    cin >> z;
    sum+=z;
    bool f=true;
    for(int i=0;i<n-1;++i){
        int x;
        cin >> x;

        if(f){
            sum+=x;
        }else{
            sum-=x;
        }f=!f;
    }cout << sum << endl;
}
